src/test/ui/impl-trait/where-allowed.rs

```diff
@@ -154,97 +155,111 @@ type InTypeAlias<R> = impl Debug;
 //~^ ERROR `impl Trait` in type aliases is unstable
 
 type InReturnInTypeAlias<R> = fn() -> impl Debug;
-//~^ ERROR `impl Trait` only allowed in function and inherent method return types
+//~^ ERROR `impl Trait` not allowed within `fn` pointer return [E0562]
 //~| ERROR `impl Trait` in type aliases is unstable
 
 // Disallowed in impl headers
 impl PartialEq<impl Debug> for () {
-    //~^ ERROR `impl Trait` only allowed in function and inherent method return types
+    //~^ ERROR `impl Trait` not allowed within trait [E0562]
 }
 
 // Disallowed in impl headers
 impl PartialEq<()> for impl Debug {
-    //~^ ERROR `impl Trait` only allowed in function and inherent method return types
+    //~^ ERROR `impl Trait` not allowed within type [E0562]
 }
 
 // Disallowed in inherent impls
 impl impl Debug {
-    //~^ ERROR `impl Trait` only allowed in function and inherent method return types
+    //~^ ERROR `impl Trait` not allowed within type [E0562]
 }
 
 // Disallowed in inherent impls
 struct InInherentImplAdt<T> { t: T }
 impl InInherentImplAdt<impl Debug> {
-    //~^ ERROR `impl Trait` only allowed in function and inherent method return types
+    //~^ ERROR `impl Trait` not allowed within type [E0562]
 }
 
-// Disallowed in where clauses
+// Allowed in where clauses
 fn in_fn_where_clause()
     where impl Debug: Debug
```

Comment:
Wait, I'm not sure this kind of bounds being allowed make much sense. What is the usecase for this?

Reply:
It's just like `fn f<T: Debug>() where T: Debug {}`.
I guess it's best to keep it this way to make it consistent.

Reply:
but unlike `fn f<T: Debug>`, given that this is a universal/"argument-position" `impl Trait`, neither the caller nor the callee can *use* this `impl Trait` (caller can't provide it, callee can't reference), so it's particularly limited.

Reply:
Yes that's a problem though.
I'll add a lint for it.
Also, what if we need a type somewhere that related to a trait, but we don't care what is the exact type, e.g. `fn f<T>(v: T) where Result<impl Default, T>: Default {}`. I think it's also a valid use-case, even though it's a by-product of impl-trait.